Implement General Framework for Unique IDs in EpiModel

AdrienLeGuillou opened the pull request:
WIP

This PR will add the "uid" attribute to the core EpiModel network functionalities. It intend to fix issues #485 and #482 

ToDo

- [X] implementation
    - [X] UID attributes
    - [X] use of UID in `set_transmat`  
- [x] unit tests
- [X] documentation

smjenness commented on R/net.accessor.R:
```diff
@@ -18,14 +17,22 @@
 #'         list object for `set_` and `add_` functions
 #'
+#' @section Core Attribute:
+#' The `append_core_attr` function initialize the attributes necessary for
```
The documentation is not built with markdown currently (I was getting an error when I tried to convert, I'll deal with a conversation later), so please change `x` to \code{x}

AdrienLeGuillou replied:
Oh ok, I updated that

AdrienLeGuillou commented:
Ready for review: 

Summary of the changes

1. one new function available to module creators: `append_core_attr()`. This function initiate the attributes of new nodes required by the rest of the EpiModel machinery (active and uid). 

2. modification of the `append_attr()` function. It can now create non existent attributes.  This makes the initial creation of nodes and the addition of new nodes in open population models more consistents. Also, it now uses the `get_attr` and `set_attr` functions internally and not specific code.

3. modification of the `set_transmat` function. It now uses UIDs instead of indexes to reference nodes

4. various corrections of bugs / typos

smjenness commented:
@AdrienLeGuillou : this is very exciting, thank you! 

Questions before I approve/merge:
1. did you test this against the user's example in #482? They were running the NME COVID model (linked in the issue) and then outputting the transmat to a phylo, and it was breaking there. 
2. I think it would be nice to see an easy demonstration of how this works with an update to this example: https://github.com/EpiModel/EpiModel-Gallery/tree/master/2018-08-SIwithVitalDynamics. Ultimately, we will need to go through all of the gallery and tutorial examples to use this new functionality, but that's a good starting point. 
3. We'll want to use this for EpiModelHIV-p/CombPrevNet, to replace the hacky version of UID there. can you just quickly scan through the code as you are doing your validation there to see if there will be any potential problem with that replacement; and then once this is merged I'll ask you to implement that replacement.

AdrienLeGuillou commented:
1. I have, I used these two files: 

- http://statnet.org/nme/d5-s4-COVIDDemog.R where I used this branch of EpiModel and added at the end

``` r
tm <- get_transmat(sim)
phylotm <- as.phylo.transmat(tm)

plot(phylotm) # will plot 38 plots
```

- http://statnet.org/nme/d5-s4-COVIDDemog-fx.R where I changed 
`dat <- append_attr(dat, item = "active", value = 1, n.new = nArrivals)` into `dat <- append_core_attr(dat, nArrivals)`

once this is approved, I could detail the necessary changes on the issue #482 

2. I made this PR implementing it: https://github.com/EpiModel/EpiModel-Gallery/pull/30 - I can make the rest of the updates pretty quickly if you want (30 min to an hour including testing) or I can wait

3. I did the update in this commit https://github.com/EpiModel/EpiModelHIV-p/commit/c69b01acacfe9bd6f81df4d7e13ecfa2f7377cf3 (the next one is just a formatting mistake) - I kept the old code in comment as I have not tested the code yet

smjenness commented:
Thanks @AdrienLeGuillou 

1. I think `append_core_attr` should also be auto-writing the other two required attributes: entrTime and exitTime. See http://statnet.org/tut/EpiModel2.html#extending-arrivals-departures. Then in the examples, it is only necessary to write append_core_attr, append_attr for status (since this may vary from model to model, it should not be in append_core_attr), and anything else bespoke. 
2. Could you move this new function out of a separate R file and into the main net accessor R file (I would like all of these utility functions together).

smjenness commented on R/net.core.attr.R:
```diff
@@ -0,0 +1,61 @@
+#'
+#' @keywords internal
+update_uids <- function(dat, n.new) {
+  last_uid <- if (is.null(dat[["_last_uid"]])) 0L else dat[["_last_uid"]]
```
Is this what initializes the `dat[["_last_uid"]]`? If not, what does? If so, shouldn't this be initialized at the current network size (e.g., `sum(dat$attr$active == 1)`) instead of `0L`?

AdrienLeGuillou replied:
It is, if `dat[["_last_uid"]]` does not exist it sets it to 0L (enforcing integer to reduce the memoy footprint)

On the first call of `append_core_attr` in the initialization module, there are no nodes yet. `n.new` are going to be created. So when `update_uids` is called for the first time, `dat[["_last_uid"]]` must be 0L as no UIDs have been used yet.

The next line creates the first `n.new` UIDs with `seq_len(n.new) + last_uid` (1:n.new in this case as `last_uid == 0L`.

`dat[["_last_uid"]]` is then updated to reflect the fact that UIDs from 1 to n.new are used

smjenness replied:
makes sense now, thanks

AdrienLeGuillou commented:
> Thanks @AdrienLeGuillou
> 
>     1. I think `append_core_attr` should also be auto-writing the other two required attributes: entrTime and exitTime. See http://statnet.org/tut/EpiModel2.html#extending-arrivals-departures. Then in the examples, it is only necessary to write append_core_attr, append_attr for status (since this may vary from model to model, it should not be in append_core_attr), and anything else bespoke.
> 
>     2. Could you move this new function out of a separate R file and into the main net accessor R file (I would like all of these utility functions together).

I agree for both, I will do the modifications. Should there also be a similar function to handle these attributes for departures?

smjenness commented:
> I agree for both, I will do the modifications. Should there also be a similar function to handle these attributes for departures?

Do you mean an auto-update accessor function specifically for the active and exitTime attributes?

AdrienLeGuillou commented:
> > I agree for both, I will do the modifications. Should there also be a similar function to handle these attributes for departures?
> 
> Do you mean an auto-update accessor function specifically for the active and exitTime attributes?

yes, mainly to be consistent in the handling of core attributes. This would also make it easier if at some point other attributes become necessary.

This way we would have `append_core_attr()` en initialization and arrival and the new function (can't think of a good name right now) in departure.

smjenness commented:
I think it's ok to leave as-is for now -- the specific calculation of exitTime is wrapped up in the broader departure function, which makes generalization tricky, and the burden is relatively low here (calling set_attr two times).

AdrienLeGuillou commented:
Ok, I just wanted to be sure

AdrienLeGuillou commented:
Changes made and checked with #482 

NB: on CombPrevNet entrTime is called arrival.time. So with the update there are both entrTime and arrival.time

I had to add an `at` argument to `append_core_attr`. At some point (not now) think I should revisit #368 as it would simplify a lot of the code and the creation of modules

smjenness commented:
> NB: on CombPrevNet entrTime is called arrival.time. So with the update there are both entrTime and arrival.time

For EpiModelHIV, please replace arrival.time with entrTime, to avoid this duplication

AdrienLeGuillou commented:
> > NB: on CombPrevNet entrTime is called arrival.time. So with the update there are both entrTime and arrival.time
> 
> For EpiModelHIV, please replace arrival.time with entrTime, to avoid this duplication

I will when I start running the combprevnet code. I don't want to break anything by changing it without running it
